### ISSUE TITLE
Parallezing `do_unmount`

### ISSUE THREAD

**Issue by anthonyryan1** [OpenRC/openrc#662]
I've been exploring shutdown times on some OpenRC servers with up to 100 hard drives connected to each machine.

I found that one of the primary bottlenecks is the sequential nature of `do_unmount` in `rc-mount.sh`. On a machine with hundreds of mount points, it will iterate through them sequentially one after another.

My first idea was to run each unmount command in the background with `&`. But that would break assumptions about unmount order, potentially creating unmount errors for people with filesystems mounted inside racing their parent to unmount.

I'm thinking the correct way to make `do_unmount` run in parallel would be to use `fstabinfo --passno ''` to get all the pass numbers, then iterate from the largest passno down to 0 using `fstabinfo --passno =N` to decide which drives can be unmounted in parallel during each batch.

I'm currently looking for feedback and concerns before creating a pull request.

**Comment by vapier** [comment]
i don't think passno works.  that assumes all mounts are in fstab to begin with, and that's def not a requirement.

this should be a compiled program so that it can actually build a graph and then process it as such using threads.

i could see there being a benefit to tickling the underlying devices in parallel in case they went to sleep -- waking back up to unmount is def slow.  i'm not sure how much overhead there actually is though if all the drives are active and we just ran umount syscall on them in serial.

**Comment by Tommimon** [comment]
I'd like to give this a try and implement a c program that uses `mountinfo` output to build a dependency graph. The program output will contain a dependency list for each mount point, those lists can of course be nested. The shell script will then run the unmount operation in parallel starting from the leaves lists and proceeding to other elements when all dependencies are met.

Do you think there's a better suitable approach for this specific issue?

**Comment by Tommimon** [comment]
To be sure I understood correctly, you suggest a `unmount_all.c` which, when executed once, will do:
1) retrieve data from `mountinfo.c`.
2) create a dependency graph on all the mount points present.
3) unmount them in parallel when possible.

If this is what you suggest, I think is a good Idea to avoid mixing C with shell when is not required.

**Comment by doot-64** [comment]
We have a proposed solution [here](https://github.com/OpenRC/openrc/pull/698) if anyone wants to take a look.

**Comment by N-R-K** [comment]
I have a WIP multi-process based parallel unmount draft. It's currently missing the fuser stuff though.

Initial impression is that the slowness you're getting is likely coming from the "unhappy" path where it fails to unmount and goes to sleep.

For testing, I used this script to create a bunch of mounts:

```
#!/bin/sh

set -- "0" "1" "2" "3" "4" "5" "6" "7" "8" "9" "A" "B" "C" "D" "E" "F"

for a in "$@"; do
	mkdir -p "/mnt/${a}"
	mount -t tmpfs -o size=256K tmpfs "/mnt/${a}"
	for b in "$@"; do
		mkdir -p "/mnt/${a}/${b}"
		mount -t tmpfs -o size=256K tmpfs "/mnt/${a}/${b}"
		for c in "$@"; do
			mkdir -p "/mnt/${a}/${b}/${c}"
			mount -t tmpfs -o size=256K tmpfs "/mnt/${a}/${b}/${c}"
			# for d in "$@"; do
			# 	mkdir -p "/mnt/${a}/${b}/${c}/${d}"
			# 	mount -t tmpfs -o size=256K tmpfs "/mnt/${a}/${b}/${c}/${d}"
			# done
		done
	done
done
```

And then tried unmounting them with 1 process at a time vs parallel and found that the parallel version was faster - but only by a small amount.

A stop-gap might be to modify the below:

https://github.com/OpenRC/openrc/blob/f247258715ade3874c196986aeff341680930f46/sh/rc-mount.sh#L72-L74

To insert another check between the sending signal and sleep; to see if there's anything else still using the mount. And if no, then avoid the sleep. Or even reduce the sleep duration, it's a heuristic anyways.

But regardless, doing parallel unmount is probably a good idea since it allows unmounting other mounts while one is waiting on a timeout. I'll see if I can finish my implementation and raise a PR sometime this week.

**Comment by Tommimon** [comment]
> I have a WIP multi-process based parallel unmount draft. It's currently missing the fuser stuff though.
> 
> Initial impression is that the slowness you're getting is likely coming from the "unhappy" path where it fails to unmount and goes to sleep.
> 
> For testing, I used this script to create a bunch of mounts:
> 
> ```
> #!/bin/sh
> 
> set -- "0" "1" "2" "3" "4" "5" "6" "7" "8" "9" "A" "B" "C" "D" "E" "F"
> 
> for a in "$@"; do
> 	mkdir -p "/mnt/${a}"
> 	mount -t tmpfs -o size=256K tmpfs "/mnt/${a}"
> 	for b in "$@"; do
> 		mkdir -p "/mnt/${a}/${b}"
> 		mount -t tmpfs -o size=256K tmpfs "/mnt/${a}/${b}"
> 		for c in "$@"; do
> 			mkdir -p "/mnt/${a}/${b}/${c}"
> 			mount -t tmpfs -o size=256K tmpfs "/mnt/${a}/${b}/${c}"
> 			# for d in "$@"; do
> 			# 	mkdir -p "/mnt/${a}/${b}/${c}/${d}"
> 			# 	mount -t tmpfs -o size=256K tmpfs "/mnt/${a}/${b}/${c}/${d}"
> 			# done
> 		done
> 	done
> done
> ```
> 
> And then tried unmounting them with 1 process at a time vs parallel and found that the parallel version was faster - but only by a small amount.
> 
> A stop-gap might be to modify the below:
> 
> [openrc/sh/rc-mount.sh](https://github.com/OpenRC/openrc/blob/f247258715ade3874c196986aeff341680930f46/sh/rc-mount.sh#L72-L74)
> 
> Lines 72 to 74 in [f247258](/OpenRC/openrc/commit/f247258715ade3874c196986aeff341680930f46)
>  fuser $f_kill$sig -k $f_opts \ 
>  	"$mnt" >/dev/null 2>&1 
>  sleep 1 
> 
> To insert another check between the sending signal and sleep; to see if there's anything else still using the mount. And if no, then avoid the sleep. Or even reduce the sleep duration, it's a heuristic anyways.
> 
> But regardless, doing parallel unmount is probably a good idea since it allows unmounting other mounts while one is waiting on a timeout. I'll see if I can finish my implementation and raise a PR sometime this week.

There is already [a working pull request](https://github.com/OpenRC/openrc/pull/698) to solve this issue. We appreciate receiving contributions and suggestions instead of duplicating work.

**Comment by N-R-K** [comment]
> There is already [a working pull request](https://github.com/OpenRC/openrc/pull/698) to solve this issue. We appreciate receiving contributions and suggestions instead of duplicating work.

Of course, for small changes and adjustments, I would just leave a review, like I already did. But as [I said earlier](https://github.com/OpenRC/openrc/pull/698#discussion_r1999180904) a multi-process based implementation is going to be *entirely* different than a multi-thread based one (see https://github.com/OpenRC/openrc/pull/811). So there's little to no duplication happening.
